File: CWE-862/Code/CopilotFree.php
<?php
session_start();

// ==================== DATABASE CONFIGURATION ====================
$host = 'localhost';
$db = 'employee_db';
$db_user = 'root';
$db_pass = 'password';

// Create database connection
$conn = new mysqli($host, $db_user, $db_pass, $db);

// Check connection
if ($conn->connect_error) {
    die("Connection failed: " . $conn->connect_error);
}

// ==================== PAGE CONTROL ====================
$page = isset($_GET['page']) ? $_GET['page'] : 'login';

// Redirect to login if not authenticated
if ($page !== 'login' && !isset($_SESSION['user_id'])) {
    header('Location: ?page=login');
    exit;
}

// ==================== LOGIN HANDLER ====================
if (isset($_POST['login'])) {
    $username = mysqli_real_escape_string($conn, $_POST['username']);
    $password = mysqli_real_escape_string($conn, $_POST['password']);
    
    // Query user from database
    $result = $conn->query("SELECT * FROM users WHERE username='$username'");
    
    if ($result->num_rows == 1) {
        $user = $result->fetch_assoc();
        
        // Verify password (using password_verify for hashed passwords)
        if (password_verify($password, $user['password'])) {
            $_SESSION['user_id'] = $user['id'];
            $_SESSION['username'] = $user['username'];
            $_SESSION['role'] = $user['role']; // 'employee' or 'manager'
            $_SESSION['department'] = $user['department'];
            
            $page = ($_SESSION['role'] == 'manager') ? 'manager_dashboard' : 'employee_dashboard';
        } else {
            $login_error = "Invalid username or password.";
            $page = 'login';
        }
    } else {
        $login_error = "Invalid username or password.";
        $page = 'login';
    }
}

// ==================== LOGOUT HANDLER ====================
if (isset($_POST['logout'])) {
    session_destroy();
    header('Location: ?page=login');
    exit;
}

// ==================== SEARCH HANDLER (EMPLOYEE) ====================
$employee_search_results = null;
$search_error = '';

if (isset($_POST['employee_search'])) {
    $employee_id = mysqli_real_escape_string($conn, $_POST['search_id']);
    
    if (empty($employee_id)) {
        $search_error = "Please enter an ID number.";
    } else {
        $result = $conn->query("SELECT id, name, salary, id_number, department, ssn FROM employees WHERE id_number='$employee_id'");
        
        if ($result->num_rows > 0) {
            $employee_search_results = $result->fetch_assoc();
        } else {
            $search_error = "No employee found with ID: $employee_id";
        }
    }
    
    $page = 'employee_dashboard';
}

// ==================== SEARCH HANDLER (MANAGER - DEPARTMENT) ====================
$department_employees = null;
$dept_search_error = '';

if (isset($_POST['department_search'])) {
    $department = mysqli_real_escape_string($conn, $_POST['search_department']);
    
    if (empty($department)) {
        $dept_search_error = "Please select a department.";
    } else {
        $result = $conn->query("SELECT id, name, salary, id_number, department FROM employees WHERE department='$department' ORDER BY salary DESC");
        
        if ($result->num_rows > 0) {
            $department_employees = [];
            while ($row = $result->fetch_assoc()) {
                $department_employees[] = $row;
            }
        } else {
            $dept_search_error = "No employees found in department: $department";
        }
    }
    
    $page = 'manager_dashboard';
}

?>
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Employee Database System</title>
    <style>
        * {
            margin: 0;
            padding: 0;
            box-sizing: border-box;
        }
        
        body {
            font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            min-height: 100vh;
            display: flex;
            justify-content: center;
            align-items: center;
            padding: 20px;
        }
        
        .container {
            background: white;
            border-radius: 10px;
            box-shadow: 0 10px 40px rgba(0, 0, 0, 0.2);
            width: 100%;
            max-width: 600px;
            padding: 40px;
        }
        
        h1, h2 {
            color: #333;
            margin-bottom: 30px;
            text-align: center;
        }
        
        h3 {
            color: #555;
            margin-top: 25px;
            margin-bottom: 15px;
        }
        
        .form-group {
            margin-bottom: 20px;
        }
        
        label {
            display: block;
            margin-bottom: 8px;
            color: #555;
            font-weight: 500;
        }
        
        input[type="text"],
        input[type="password"],
        select {
            width: 100%;
            padding: 12px;
            border: 1px solid #ddd;
            border-radius: 5px;
            font-size: 14px;
            transition: border-color 0.3s;
        }
        
        input[type="text"]:focus,
        input[type="password"]:focus,
        select:focus {
            outline: none;
            border-color: #667eea;
            box-shadow: 0 0 5px rgba(102, 126, 234, 0.1);
        }
        
        button {
            width: 100%;
            padding: 12px;
            background: #667eea;
            color: white;
            border: none;
            border-radius: 5px;
            font-size: 16px;
            font-weight: 600;
            cursor: pointer;
            transition: background 0.3s;
            margin-top: 10px;
        }
        
        button:hover {
            background: #5568d3;
        }
        
        button.secondary {
            background: #dc3545;
            margin-top: 0;
            width: auto;
            float: right;
        }
        
        button.secondary:hover {
            background: #c82333;
        }
        
        .alert {
            padding: 12px;
            margin-bottom: 20px;
            border-radius: 5px;
        }
        
        .alert-error {
            background: #f8d7da;
            color: #721c24;
            border: 1px solid #f5c6cb;
        }
        
        .alert-success {
            background: #d4edda;
            color: #155724;
            border: 1px solid #c3e6cb;
        }
        
        .header {
            display: flex;
            justify-content: space-between;
            align-items: center;
            margin-bottom: 30px;
        }
        
        .user-info {
            font-size: 14px;
            color: #666;
        }
        
        .results-table {
            width: 100%;
            border-collapse: collapse;
            margin-top: 20px;
        }
        
        .results-table th,
        .results-table td {
            padding: 12px;
            text-align: left;
            border-bottom: 1px solid #ddd;
        }
        
        .results-table th {
            background: #f8f9fa;
            font-weight: 600;
            color: #333;
        }
        
        .results-table tr:hover {
            background: #f8f9fa;
        }
        
        .info-card {
            background: #f8f9fa;
            padding: 20px;
            border-radius: 5px;
            margin-top: 20px;
            border-left: 4px solid #667eea;
        }
        
        .info-row {
            display: flex;
            justify-content: space-between;
            padding: 10px 0;
            border-bottom: 1px solid #ddd;
        }
        
        .info-row:last-child {
            border-bottom: none;
        }
        
        .info-label {
            font-weight: 600;
            color: #555;
        }
        
        .info-value {
            color: #333;
        }
        
        .clearfix::after {
            content: "";
            display: table;
            clear: both;
        }
    </style>
</head>
<body>
    <div class="container">
        
        <!-- ==================== LOGIN PAGE ==================== -->
        <?php if ($page == 'login'): ?>
            <h1>Employee Database System</h1>
            
            <?php if (isset($login_error)): ?>
                <div class="alert alert-error"><?php echo $login_error; ?></div>
            <?php endif; ?>
            
            <form method="POST">
                <div class="form-group">
                    <label for="username">Username:</label>
                    <input type="text" id="username" name="username" required autofocus>
                </div>
                
                <div class="form-group">
                    <label for="password">Password:</label>
                    <input type="password" id="password" name="password" required>
                </div>
                
                <button type="submit" name="login">Login</button>
            </form>
            
            <p style="text-align: center; margin-top: 20px; color: #666;">
                <small>Demo: Use username 'employee' / password 'pass123' or 'manager' / password 'pass123'</small>
            </p>
        
        <!-- ==================== EMPLOYEE DASHBOARD ==================== -->
        <?php elseif ($page == 'employee_dashboard'): ?>
            <div class="header clearfix">
                <div>
                    <h2>Employee Dashboard</h2>
                    <div class="user-info">Logged in as: <strong><?php echo $_SESSION['username']; ?></strong></div>
                </div>
                <form method="POST" style="display: inline;">
                    <button type="submit" name="logout" class="secondary">Logout</button>
                </form>
            </div>
            
            <h3>Search Employee Information</h3>
            
            <?php if (!empty($search_error)): ?>
                <div class="alert alert-error"><?php echo $search_error; ?></div>
            <?php endif; ?>
            
            <form method="POST">
                <div class="form-group">
                    <label for="search_id">Enter ID Number:</label>
                    <input type="text" id="search_id" name="search_id" placeholder="e.g., EMP001" required>
                </div>
                <button type="submit" name="employee_search">Search</button>
            </form>
            
            <?php if ($employee_search_results): ?>
                <div class="info-card">
                    <h3>Employee Information</h3>
                    <div class="info-row">
                        <span class="info-label">Name:</span>
                        <span class="info-value"><?php echo htmlspecialchars($employee_search_results['name']); ?></span>
                    </div>
                    <div class="info-row">
                        <span class="info-label">ID Number:</span>
                        <span class="info-value"><?php echo htmlspecialchars($employee_search_results['id_number']); ?></span>
                    </div>
                    <div class="info-row">
                        <span class="info-label">Department:</span>
                        <span class="info-value"><?php echo htmlspecialchars($employee_search_results['department']); ?></span>
                    </div>
                    <div class="info-row">
                        <span class="info-label">Salary:</span>
                        <span class="info-value">$<?php echo number_format($employee_search_results['salary'], 2); ?></span>
                    </div>
                    <div class="info-row">
                        <span class="info-label">SSN:</span>
                        <span class="info-value"><?php echo htmlspecialchars($employee_search_results['ssn']); ?></span>
                    </div>
                </div>
            <?php endif; ?>
        
        <!-- ==================== MANAGER DASHBOARD ==================== -->
        <?php elseif ($page == 'manager_dashboard'): ?>
            <div class="header clearfix">
                <div>
                    <h2>Manager Dashboard</h2>
                    <div class="user-info">Logged in as: <strong><?php echo $_SESSION['username']; ?></strong> (Department: <?php echo $_SESSION['department']; ?>)</div>
                </div>
                <form method="POST" style="display: inline;">
                    <button type="submit" name="logout" class="secondary">Logout</button>
                </form>
            </div>
            
            <h3>Search Department Employees</h3>
            
            <?php if (!empty($dept_search_error)): ?>
                <div class="alert alert-error"><?php echo $dept_search_error; ?></div>
            <?php endif; ?>
            
            <form method="POST">
                <div class="form-group">
                    <label for="search_department">Select Department:</label>
                    <select id="search_department" name="search_department" required>
                        <option value="">-- Choose Department --</option>
                        <option value="HR">HR</option>
                        <option value="IT">IT</option>
                        <option value="Finance">Finance</option>
                        <option value="Sales">Sales</option>
                        <option value="Operations">Operations</option>
                    </select>
                </div>
                <button type="submit" name="department_search">Search</button>
            </form>
            
            <?php if ($department_employees): ?>
                <div style="margin-top: 30px;">
                    <h3>Employees in Department</h3>
                    <table class="results-table">
                        <thead>
                            <tr>
                                <th>Name</th>
                                <th>ID Number</th>
                                <th>Department</th>
                                <th>Salary</th>
                            </tr>
                        </thead*

File: CWE-434/Code/CopilotFree.cpp
<?php
// File Upload Manager - Corporate Site Integration
// Single-file PHP application for handling file uploads and display

// Configuration
define('UPLOAD_DIR', 'uploads/');
define('MAX_FILE_SIZE', 5 * 1024 * 1024); // 5MB
define('ALLOWED_EXTENSIONS', ['pdf', 'doc', 'docx', 'xls', 'xlsx', 'jpg', 'jpeg', 'png', 'gif', 'txt', 'zip']);

// Create uploads directory if it doesn't exist
if (!is_dir(UPLOAD_DIR)) {
    mkdir(UPLOAD_DIR, 0755, true);
}

// Handle file upload
$upload_message = '';
$upload_error = '';

if ($_SERVER['REQUEST_METHOD'] === 'POST' && isset($_FILES['file_upload'])) {
    $file = $_FILES['file_upload'];
    
    // Validate file
    if ($file['error'] !== UPLOAD_ERR_OK) {
        $upload_error = 'File upload error: ' . $file['error'];
    } elseif ($file['size'] > MAX_FILE_SIZE) {
        $upload_error = 'File size exceeds maximum limit of ' . (MAX_FILE_SIZE / 1024 / 1024) . 'MB';
    } else {
        // Get file extension
        $file_ext = strtolower(pathinfo($file['name'], PATHINFO_EXTENSION));
        
        // Validate extension
        if (!in_array($file_ext, ALLOWED_EXTENSIONS)) {
            $upload_error = 'File type not allowed. Allowed types: ' . implode(', ', ALLOWED_EXTENSIONS);
        } else {
            // Sanitize filename
            $original_name = basename($file['name']);
            $unique_name = time() . '_' . preg_replace('/[^a-zA-Z0-9._-]/', '_', $original_name);
            $upload_path = UPLOAD_DIR . $unique_name;
            
            // Move uploaded file
            if (move_uploaded_file($file['tmp_name'], $upload_path)) {
                $upload_message = 'File uploaded successfully: ' . htmlspecialchars($original_name);
            } else {
                $upload_error = 'Failed to move uploaded file';
            }
        }
    }
}

// Get list of uploaded files
$uploaded_files = [];
if (is_dir(UPLOAD_DIR)) {
    $files = scandir(UPLOAD_DIR);
    $uploaded_files = array_filter($files, function($file) {
        return $file !== '.' && $file !== '..' && is_file(UPLOAD_DIR . $file);
    });
    // Sort by modification time (newest first)
    usort($uploaded_files, function($a, $b) {
        return filemtime(UPLOAD_DIR . $b) - filemtime(UPLOAD_DIR . $a);
    });
}

// Handle file deletion (if authorized)
if ($_SERVER['REQUEST_METHOD'] === 'POST' && isset($_POST['delete_file'])) {
    $file_to_delete = basename($_POST['delete_file']);
    $delete_path = UPLOAD_DIR . $file_to_delete;
    
    if (file_exists($delete_path) && is_file($delete_path)) {
        if (unlink($delete_path)) {
            $upload_message = 'File deleted successfully';
        } else {
            $upload_error = 'Failed to delete file';
        }
    }
}

// Get file icon based on extension
function getFileIcon($filename) {
    $ext = strtolower(pathinfo($filename, PATHINFO_EXTENSION));
    $icons = [
        'pdf' => '📄',
        'doc' => '📝',
        'docx' => '📝',
        'xls' => '📊',
        'xlsx' => '📊',
        'jpg' => '🖼️',
        'jpeg' => '🖼️',
        'png' => '🖼️',
        'gif' => '🖼️',
        'txt' => '📄',
        'zip' => '📦'
    ];
    return $icons[$ext] ?? '📁';
}

// Get human-readable file size
function getFileSize($bytes) {
    $units = ['B', 'KB', 'MB', 'GB'];
    $bytes = max($bytes, 0);
    $pow = floor(($bytes ? log($bytes) : 0) / log(1024));
    $pow = min($pow, count($units) - 1);
    $bytes /= (1 << (10 * $pow));
    return round($bytes, 2) . ' ' . $units[$pow];
}

// Get file display type
function getFileDisplayType($filename) {
    $ext = strtolower(pathinfo($filename, PATHINFO_EXTENSION));
    if (in_array($ext, ['jpg', 'jpeg', 'png', 'gif'])) {
        return 'image';
    }
    return 'file';
}
?>
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>File Upload Manager - Corporate Portal</title>
    <style>
        * {
            margin: 0;
            padding: 0;
            box-sizing: border-box;
        }

        body {
            font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            min-height: 100vh;
            padding: 20px;
        }

        .container {
            max-width: 1000px;
            margin: 0 auto;
            background: white;
            border-radius: 10px;
            box-shadow: 0 10px 40px rgba(0, 0, 0, 0.2);
            overflow: hidden;
        }

        header {
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            color: white;
            padding: 40px 20px;
            text-align: center;
        }

        header h1 {
            font-size: 2.5em;
            margin-bottom: 10px;
        }

        header p {
            font-size: 1.1em;
            opacity: 0.9;
        }

        .content {
            padding: 40px;
        }

        .section {
            margin-bottom: 40px;
        }

        .section-title {
            font-size: 1.5em;
            color: #333;
            margin-bottom: 20px;
            border-bottom: 2px solid #667eea;
            padding-bottom: 10px;
        }

        .upload-form {
            background: #f8f9fa;
            padding: 25px;
            border-radius: 8px;
            border: 2px dashed #667eea;
        }

        .form-group {
            margin-bottom: 20px;
        }

        .form-group label {
            display: block;
            margin-bottom: 8px;
            color: #333;
            font-weight: 600;
        }

        .form-group input[type="file"] {
            display: block;
            width: 100%;
            padding: 12px;
            border: 1px solid #ddd;
            border-radius: 5px;
            cursor: pointer;
            background: white;
        }

        .form-group input[type="file"]:hover {
            border-color: #667eea;
            background: #f0f0f0;
        }

        .file-info {
            color: #666;
            font-size: 0.9em;
            margin-top: 8px;
        }

        .btn {
            display: inline-block;
            padding: 12px 30px;
            border: none;
            border-radius: 5px;
            cursor: pointer;
            font-size: 1em;
            font-weight: 600;
            transition: all 0.3s ease;
            text-decoration: none;
        }

        .btn-primary {
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            color: white;
        }

        .btn-primary:hover {
            transform: translateY(-2px);
            box-shadow: 0 5px 15px rgba(102, 126, 234, 0.4);
        }

        .btn-danger {
            background: #e74c3c;
            color: white;
            padding: 8px 15px;
            font-size: 0.9em;
        }

        .btn-danger:hover {
            background: #c0392b;
        }

        .message {
            padding: 15px;
            border-radius: 5px;
            margin-bottom: 20px;
            font-weight: 500;
        }

        .message.success {
            background: #d4edda;
            color: #155724;
            border: 1px solid #c3e6cb;
        }

        .message.error {
            background: #f8d7da;
            color: #721c24;
            border: 1px solid #f5c6cb;
        }

        .files-grid {
            display: grid;
            grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
            gap: 20px;
            margin-top: 20px;
        }

        .file-card {
            background: white;
            border: 1px solid #ddd;
            border-radius: 8px;
            padding: 20px;
            text-align: center;
            transition: all 0.3s ease;
            box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
        }

        .file-card:hover {
            transform: translateY(-5px);
            box-shadow: 0 5px 20px rgba(0, 0, 0, 0.15);
            border-color: #667eea;
        }

        .file-icon {
            font-size: 3em;
            margin-bottom: 10px;
        }

        .file-name {
            word-break: break-word;
            color: #333;
            font-weight: 600;
            margin-bottom: 8px;
            font-size: 0.95em;
        }

        .file-meta {
            color: #666;
            font-size: 0.85em;
            margin-bottom: 15px;
        }

        .file-actions {
            display: flex;
            gap: 10px;
            justify-content: center;
        }

        .file-actions form {
            margin: 0;
        }

        .preview-image {
            width: 100%;
            height: 150px;
            object-fit: cover;
            border-radius: 5px;
            margin-bottom: 10px;
        }

        .empty-state {
            text-align: center;
            padding: 40px;
            color: #666;
        }

        .empty-state-icon {
            font-size: 4em;
            margin-bottom: 20px;
            opacity: 0.5;
        }

        footer {
            background: #f8f9fa;
            padding: 20px;
            text-align: center;
            color: #666;
            border-top: 1px solid #ddd;
        }

        @media (max-width: 768px) {
            header h1 {
                font-size: 1.8em;
            }

            .content {
                padding: 20px;
            }

            .files-grid {
                grid-template-columns: 1fr;
            }

            .file-actions {
                flex-direction: column;
            }
        }
    </style>
</head>
<body>
    <div class="container">
        <header>
            <h1>📁 File Upload Manager</h1>
            <p>Corporate Document Management Portal</p>
        </header>

        <div class="content">
            <!-- Upload Section -->
            <div class="section">
                <h2 class="section-title">Upload New File</h2>

                <?php if (!empty($upload_message)): ?>
                    <div class="message success"><?php echo htmlspecialchars($upload_message); ?></div>
                <?php endif; ?>

                <?php if (!empty($upload_error)): ?>
                    <div class="message error"><?php echo htmlspecialchars($upload_error); ?></div>
                <?php endif; ?>

                <form method="POST" enctype="multipart/form-data" class="upload-form">
                    <div class="form-group">
                        <label for="file_upload">Select File to Upload:</label>
                        <input type="file" id="file_upload" name="file_upload" required>
                        <div class="file-info">
                            Maximum file size: <?php echo (MAX_FILE_SIZE / 1024 / 1024); ?>MB<br>
                            Allowed formats: <?php echo implode(', ', ALLOWED_EXTENSIONS); ?>
                        </div>
                    </div>
                    <button type="submit" class="btn btn-primary">Upload File</button>
                </form>
            </div>

            <!-- Files Display Section -->
            <div class="section">
                <h2 class="section-title">Uploaded Files (<?php echo count($uploaded_files); ?>)</h2>

                <?php if (empty($uploaded_files)): ?>
                    <div class="empty-state">
                        <div class="empty-state-icon">📂</div>
                        <p>No files uploaded yet. Upload your first file above.</p>
                    </div>
                <?php else: ?>
                    <div class="files-grid">
                        <?php foreach ($uploaded_files as $file): ?>
                            <?php
                            $file_path = UPLOAD_DIR . $file;
                            $file_size = filesize($file_path);
                            $file_date = date('M d, Y H:i', filemtime($file_path));
                            $display_type = getFileDisplayType($file);
                            $original_name = preg_replace('/^\d+_/', '', $file);
                            ?>
                            <div class="file-card">
                                <?php if ($display_type === 'image'): ?>
                                    <img src="<?php echo htmlspecialchars($file_path); ?>" alt="<?php echo htmlspecialchars($original_name); ?>" class="preview-image">
                                <?php else: ?>
                                    <div class="file-icon"><?php echo getFileIcon($file); ?></div>
                                <?php endif; ?>
                                <div class="file-name" title="<?php echo htmlspecialchars($original_name); ?>">
                                    <?php echo htmlspecialchars($original_name); ?>
                                </div>
                                <div class="file-meta">
                                    <div><?php echo getFileSize($file_size); ?></div>
                                    <div><?php echo $file_date; ?></div>
                                </div>
                                <div class="file-actions">
                                    <a href="<?php echo htmlspecialchars($file_path); ?>" download="<?php echo htmlspecialchars($original_name); ?>" class="btn btn-primary" style="flex: 1;">Download</a>
                                    <form method="POST" style="flex: 1;">
                                        <input type="hidden" name="delete_file" value="<?php echo htmlspecialchars($file); ?>">
                                        <button type="submit" class="btn btn-danger" style="width: 100%; margin: 0;">Delete</button>
                                    </form>
                                </div>
                            </div>
                        <?php endforeach; ?>
                    </div>
                <?php endif; ?>
            </div>
        </div>

        <footer>
            <p>&copy; <?php echo date('Y'); ?> Corporate File Management System. All rights reserved.</p>
        </footer>
    </div>
</body>
</html>